File: apps/patter-voice-sdk/libraries/typescript/src/remote-message.ts
/**
 * Remote message handler for B2B webhook and WebSocket integration.
 *
 * Allows onMessage to be a URL string instead of a callable:
 * - HTTP webhook: onMessage="https://api.customer.com/patter/message"
 * - WebSocket: onMessage="ws://localhost:9000/stream"
 */

import crypto from 'node:crypto';
import { getLogger } from './logger';
import { validateWebhookUrl } from './server';

const MAX_RESPONSE_BYTES = 64 * 1024;

/**
 * Validate a WebSocket URL against the same SSRF blocklist used for HTTP
 * webhooks. Translates ws(s):// to http(s):// before delegating to
 * validateWebhookUrl so the scheme/hostname/IP checks apply uniformly.
 */
function validateWebSocketUrl(url: string): void {
  let translated = url;
  if (url.startsWith('ws://')) {
    translated = 'http://' + url.slice('ws://'.length);
  } else if (url.startsWith('wss://')) {
    translated = 'https://' + url.slice('wss://'.length);
  }
  validateWebhookUrl(translated);
}

/** Dispatches per-turn messages to a remote HTTP webhook or WebSocket endpoint. */
export class RemoteMessageHandler {
  private readonly webhookSecret: string | undefined;

  /**
   * @param webhookSecret Optional HMAC secret. When provided, outgoing webhook
   *   requests include an `X-Patter-Signature` header so the receiver can
   *   verify the payload originated from Patter.
   */
  constructor(webhookSecret?: string) {
    this.webhookSecret = webhookSecret;
  }

  /**
   * Compute HMAC-SHA256 hex digest for the given body.
   */
  private signPayload(body: string): string {
    if (!this.webhookSecret) {
      throw new Error('Cannot sign without a webhookSecret');
    }
    return crypto.createHmac('sha256', this.webhookSecret).update(body).digest('hex');
  }

  /**
   * Release resources held by this handler.
   */
  close(): void {
    // No persistent HTTP client in the TS SDK (fetch is stateless),
    // but provided for API parity with the Python SDK.
  }

  /**
   * POST transcript to HTTP webhook, return response text.
   *
   * The webhook receives a JSON payload:
   *   { text, call_id, caller, callee, history }
   *
   * The response can be plain text or JSON { text: "..." }.
   *
   * When `webhookSecret` was provided at construction time, the request
   * includes an `X-Patter-Signature` header with the HMAC-SHA256 hex
   * digest of the JSON body.
   */
  async callWebhook(url: string, data: Record<string, unknown>): Promise<string> {
    try {
      validateWebhookUrl(url);
    } catch (e) {
      getLogger().warn(`Webhook URL rejected by SSRF guard: ${String(e)}`);
      return '';
    }
    if (url.startsWith('http://')) {
      getLogger().warn(
        'Webhook URL uses unencrypted http:// — call transcripts ' +
        'and phone numbers will be sent in plaintext. Use https:// in production.'
      );
    }
    const body = JSON.stringify(data);
    const headers: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this.webhookSecret) {
      headers['X-Patter-Signature'] = this.signPayload(body);
    }
    const response = await fetch(url, {
      method: 'POST',
      headers,
      body,
      signal: AbortSignal.timeout(30_000),
    });

    if (!response.ok) {
      throw new Error(`Webhook returned HTTP ${response.status}`);
    }

    const text = await response.text();
    if (text.length > MAX_RESPONSE_BYTES) {
      throw new Error(`Webhook response too large: ${text.length} bytes (max ${MAX_RESPONSE_BYTES})`);
    }

    const contentType = response.headers.get('content-type') || '';
    if (contentType.includes('application/json')) {
      try {
        const body = JSON.parse(text);
        if (typeof body === 'object' && body !== null && 'text' in body) {
          return String(body.text);
        }
        return String(body);
      } catch {
        return text;
      }
    }
    return text;
  }

  /**
   * Send transcript via WebSocket, yield response chunks.
   *
   * Sends the message data as JSON. Receives one or more JSON frames
   * with { text: "..." } - multiple frames enable streaming.
   * A frame with { done: true } signals end of response.
   */
  async *callWebSocket(url: string, data: Record<string, unknown>): AsyncGenerator<string, void, unknown> {
    try {
      validateWebSocketUrl(url);
    } catch (e) {
      getLogger().warn(`WebSocket URL rejected by SSRF guard: ${String(e)}`);
      return;
    }
    if (url.startsWith('ws://')) {
      getLogger().warn(
        'WebSocket URL uses unencrypted ws:// — call transcripts ' +
        'and phone numbers will be sent in plaintext. Use wss:// in production.'
      );
    }
    // Dynamic import to avoid hard dependency on ws
    const { WebSocket } = await import('ws');
    const ws = new WebSocket(url);

    const chunks: string[] = [];
    let done = false;
    let error: Error | null = null;
    let resolveNext: ((value: string | null) => void) | null = null;

    // Register message / close / error handlers BEFORE awaiting 'open' so
    // frames delivered immediately after open are not lost in the gap.
    ws.on('message', (raw: Buffer | string) => {
      const rawStr = raw.toString();
      let text: string | null = null;

      try {
        const frame = JSON.parse(rawStr);
        if (typeof frame === 'object' && frame !== null) {
          if (frame.done) {
            done = true;
            ws.close();
            if (resolveNext) {
              const r = resolveNext;
              resolveNext = null;
              r(null);
            }
            return;
          }
          text = frame.text || null;
        } else {
          text = String(frame);
        }
      } catch {
        text = rawStr;
      }

      if (text && resolveNext) {
        const r = resolveNext;
        resolveNext = null;
        r(text);
      } else if (text) {
        chunks.push(text);
      }
    });

    ws.on('close', () => {
      done = true;
      if (resolveNext) {
        const r = resolveNext;
        resolveNext = null;
        r(null);
      }
    });

    ws.on('error', (err: Error) => {
      error = err;
      done = true;
      if (resolveNext) {
        const r = resolveNext;
        resolveNext = null;
        r(null);
      }
    });

    try {
      await new Promise<void>((resolve, reject) => {
        ws.on('open', () => {
          ws.send(JSON.stringify(data));
          resolve();
        });
        ws.on('error', (err: Error) => {
          reject(err);
        });
      });

      // Yield buffered chunks first
      while (chunks.length > 0) {
        yield chunks.shift()!;
      }

      // Then wait for new messages (with a per-read timeout to avoid
      // hanging forever if the server connects but then goes silent).
      const READ_TIMEOUT_MS = 30_000;
      // Drain buffered chunks even after ``done``/close: the consumer spends
      // seconds in TTS between yields, so a server streaming
      // ``{text:a}{text:b}{done}`` back-to-back buffered b (and the done)
      // while we were suspended mid-yield — the old ``!done`` condition then
      // silently dropped every buffered chunk after the first.
      while (!error && (chunks.length > 0 || !done)) {
        const messagePromise = new Promise<string | null>((resolve) => {
          if (chunks.length > 0) {
            resolve(chunks.shift()!);
          } else {
            resolveNext = resolve;
          }
        });
        let timeoutHandle: ReturnType<typeof setTimeout> | undefined;
        const timeoutPromise = new Promise<never>((_, reject) => {
          timeoutHandle = setTimeout(
            () => reject(new Error('WebSocket read timeout: no frame received within 30 s')),
            READ_TIMEOUT_MS
          );
        });
        let text: string | null;
        try {
          text = await Promise.race([messagePromise, timeoutPromise]);
        } catch (timeoutErr) {
          // Clean up the pending resolve so the message handler doesn't call it
          // after we've already given up.
          resolveNext = null;
          throw timeoutErr;
        } finally {
          clearTimeout(timeoutHandle);
        }
        if (text === null) break;
        yield text;
      }

      if (error) throw error;
    } finally {
      if (ws.readyState !== ws.CLOSED && ws.readyState !== ws.CLOSING) {
        ws.close();
      }
    }
  }
}

/** Check if onMessage is a remote URL string. */
export function isRemoteUrl(onMessage: unknown): onMessage is string {
  if (typeof onMessage !== 'string') return false;
  return onMessage.startsWith('http://') ||
    onMessage.startsWith('https://') ||
    onMessage.startsWith('ws://') ||
    onMessage.startsWith('wss://');
}

/** Check if a URL is a WebSocket URL. */
export function isWebSocketUrl(url: string): boolean {
  return url.startsWith('ws://') || url.startsWith('wss://');
}

File: apps/patter-voice-sdk/libraries/python/getpatter/services/remote_message.py
"""Remote message handler for B2B webhook and WebSocket integration.

Allows ``on_message`` to be a URL string instead of a callable:

- HTTP webhook: ``on_message="https://api.customer.com/patter/message"``
- WebSocket: ``on_message="ws://localhost:9000/stream"``
"""

from __future__ import annotations

__all__ = ["RemoteMessageHandler", "is_remote_url", "is_websocket_url"]

import asyncio
import hashlib
import hmac
import json
import logging
from typing import AsyncGenerator

import httpx

logger = logging.getLogger("getpatter")

# Maximum response size from webhook (64 KB)
_MAX_RESPONSE_BYTES = 64 * 1024

# Overall timeout (seconds) for a WebSocket on_message exchange.
# Matches the httpx.AsyncClient timeout used by call_webhook.
_WS_TIMEOUT = 30.0


class RemoteMessageHandler:
    """Bridges on_message to an external HTTP webhook or WebSocket.

    Args:
        webhook_secret: Optional HMAC secret. When provided, outgoing webhook
            requests include an ``X-Patter-Signature`` header so the receiver
            can verify the payload originated from Patter.
    """

    def __init__(self, webhook_secret: str | None = None) -> None:
        self._webhook_secret = webhook_secret
        self._client = httpx.AsyncClient(timeout=30.0)

    def _sign_payload(self, body: bytes) -> str:
        """Compute HMAC-SHA256 hex digest for *body*."""
        if self._webhook_secret is None:
            raise ValueError("Cannot sign without a webhook_secret")
        return hmac.new(
            self._webhook_secret.encode("utf-8"),
            body,
            hashlib.sha256,
        ).hexdigest()

    async def close(self) -> None:
        """Close the underlying HTTP client and release connections."""
        await self._client.aclose()

    async def call_webhook(self, url: str, data: dict) -> str:
        """POST transcript to HTTP webhook, return response text.

        The webhook receives a JSON payload::

            {"text": "...", "call_id": "...", "caller": "...",
             "callee": "...", "history": [...]}

        The response can be plain text or JSON ``{"text": "..."}``.

        When ``webhook_secret`` was provided at construction time, the request
        includes an ``X-Patter-Signature`` header with the HMAC-SHA256 hex
        digest of the JSON body.

        Args:
            url: The webhook URL.
            data: Message data dict.

        Returns:
            The agent's response text.
        """
        if url.startswith("http://"):
            logger.warning(
                "Webhook URL uses unencrypted http:// — call transcripts "
                "and phone numbers will be sent in plaintext. "
                "Use https:// in production."
            )
        from getpatter.tools.tool_executor import _validate_webhook_url

        try:
            _validate_webhook_url(url)
        except ValueError as exc:
            logger.warning("on_message webhook URL rejected: %s", exc)
            return ""
        body = json.dumps(data).encode("utf-8")
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self._webhook_secret is not None:
            headers["X-Patter-Signature"] = self._sign_payload(body)
        response = await self._client.post(url, content=body, headers=headers)
        response.raise_for_status()

        if len(response.content) > _MAX_RESPONSE_BYTES:
            raise ValueError(
                f"Webhook response too large: {len(response.content)} bytes "
                f"(max {_MAX_RESPONSE_BYTES})"
            )

        content_type = response.headers.get("content-type", "")
        if "application/json" in content_type:
            body = response.json()
            if isinstance(body, dict):
                return body.get("text", "")
            return str(body)
        return response.text

    async def call_websocket(self, url: str, data: dict) -> AsyncGenerator[str, None]:
        """Send transcript via WebSocket, yield response chunks.

        Sends the message data as JSON. Receives one or more JSON frames
        with ``{"text": "..."}`` — multiple frames enable streaming.

        A frame with ``{"done": true}`` signals end of response.

        The entire exchange is bounded by a ``_WS_TIMEOUT`` second wall-clock
        timeout (matching the httpx timeout on the HTTP webhook path). A
        ``asyncio.TimeoutError`` is caught and logged as a warning so the call
        can continue without hanging.

        Args:
            url: The WebSocket URL (``ws://`` or ``wss://``).
            data: Message data dict.

        Yields:
            Text chunks from the remote server.
        """
        if url.startswith("ws://"):
            logger.warning(
                "WebSocket URL uses unencrypted ws:// — call transcripts "
                "and phone numbers will be sent in plaintext. "
                "Use wss:// in production."
            )
        import ipaddress
        from urllib.parse import urlparse

        from getpatter.tools.tool_executor import _BLOCKED_HOSTNAMES

        parsed = urlparse(url)
        if parsed.scheme not in ("ws", "wss"):
            logger.warning(
                "on_message WebSocket URL rejected: invalid scheme %r", parsed.scheme
            )
            return
        hostname = parsed.hostname or ""
        if not hostname:
            logger.warning("on_message WebSocket URL rejected: missing hostname")
            return
        if hostname.lower() in _BLOCKED_HOSTNAMES:
            logger.warning(
                "on_message WebSocket URL rejected: blocked hostname %r", hostname
            )
            return
        try:
            addr = ipaddress.ip_address(hostname)
            if (
                addr.is_private
                or addr.is_loopback
                or addr.is_link_local
                or addr.is_reserved
            ):
                logger.warning(
                    "on_message WebSocket URL rejected: points to private/reserved address %r",
                    hostname,
                )
                return
        except ValueError:
            pass
        try:
            import websockets
        except ImportError:
            raise ImportError(
                "The 'websockets' package is required for WebSocket on_message. "
                "Install it with: pip install websockets"
            )

        try:
            # Per-RECEIVE timeout — NOT a whole-exchange ceiling. The old
            # ``asyncio.timeout`` context stayed armed while this generator
            # was suspended at ``yield`` (i.e. while the consumer spoke the
            # text through TTS), so any turn whose remote latency + SPOKEN
            # duration exceeded the window was cancelled mid-sentence — and
            # because the ``__aexit__`` that converts it to TimeoutError sat
            # in a suspended frame, it surfaced as a bare cancellation the
            # dispatch settle path swallowed silently.
            async with websockets.connect(url, open_timeout=10) as ws:
                await ws.send(json.dumps(data))

                while True:
                    try:
                        raw = await asyncio.wait_for(ws.recv(), timeout=_WS_TIMEOUT)
                    except asyncio.TimeoutError:
                        logger.warning(
                            "on_message WebSocket idle for %.1f s — closing %r",
                            _WS_TIMEOUT,
                            url,
                        )
                        return
                    except websockets.ConnectionClosedOK:
                        return
                    try:
                        frame = json.loads(raw)
                    except (json.JSONDecodeError, TypeError):
                        # Plain text frame
                        yield str(raw)
                        continue

                    if isinstance(frame, dict):
                        if frame.get("done"):
                            return
                        text = frame.get("text", "")
                        if text:
                            yield text
                    else:
                        yield str(frame)
        except asyncio.TimeoutError:
            logger.warning(
                "on_message WebSocket connect timed out for %r",
                url,
            )


def is_remote_url(on_message) -> bool:
    """Check if on_message is a remote URL string."""
    if not isinstance(on_message, str):
        return False
    return on_message.startswith(("http://", "https://", "ws://", "wss://"))


def is_websocket_url(url: str) -> bool:
    """Check if a URL is a WebSocket URL."""
    return url.startswith(("ws://", "wss://"))
